File: src/controllers/handlers-options.ts
// Constants
import { REPOSITORY } from '@/constants/github'

// Third-Party libraries
import opener from 'opener'

// Utils
import { generateCommitlintConfig } from '@/utils/commitlint'
import { writeMessage } from '@/utils/console'
import { getErrorMessage } from '@/utils/errors'
import { generateHuskyConfig } from '@/utils/husky-library'
import { exists } from '@/utils/user-os'

// NodeJS
import process from 'node:process'

// User-Input
import { addCommitlint } from '@/user-input/add-commitlint'
import { shouldPublishToNPM } from '@/user-input/npm'
import { getPackageManger } from '@/user-input/package-managers'

export const handlerOptionBuild = async () => {
  try {
    let shouldPublishToNpm = false

    const packageJsonPath = `${process.cwd()}/package.json`

    const existPackageJsonInTheCurrentDirectory = await exists(packageJsonPath)

    if (!existPackageJsonInTheCurrentDirectory) {
      writeMessage({
        type: 'error',
        message: getErrorMessage('NotFound')
      })

      process.exit(1)
    }

    const packageManagerToUse = await getPackageManger()

    if (packageManagerToUse === 'yarn') {
      shouldPublishToNpm = await shouldPublishToNPM()
    }

    const useCommitlint = await addCommitlint()

    await generateHuskyConfig({
      packageManagerToUse,
      packageJsonPath,
      useCommitlint,
      shouldPublishToNpm
    })

    if (useCommitlint) {
      await generateCommitlintConfig({ packageManagerToUse, packageJsonPath })
    }
  } catch {
    writeMessage({
      type: 'error',
      message: getErrorMessage('Default')
    })
    process.exit(1)
  }
}

export const handlerOptionCollaborate = () => {
  writeMessage({
    type: 'info',
    message: 'Opening the GitHub repository...'
  })

  setTimeout(() => opener(REPOSITORY), 500)
}

File: internal/handlers/handlers_options.go
package handlers

import (
	"os"
	"sync"

	"github.com/RaulCatalinas/HuskyBC/internal/constants"
	"github.com/RaulCatalinas/HuskyBC/internal/enums"
	errorMessages "github.com/RaulCatalinas/HuskyBC/internal/error_messages"
	userInput "github.com/RaulCatalinas/HuskyBC/internal/user_input"
	"github.com/RaulCatalinas/HuskyBC/internal/utils"
	"github.com/toqueteos/webbrowser"
)

func HandlerOptionCollaborate() {
	utils.WriteMessage(utils.WriteMessageProps{
		Type:    enums.MessageTypeInfo,
		Message: "Opening the GitHub repository...",
	})

	if err := webbrowser.Open(constants.REPOSITORY); err != nil {
		utils.WriteMessage(utils.WriteMessageProps{
			Type:    enums.MessageTypeError,
			Message: errorMessages.PROCESS_ERROR_MESSAGES[enums.GitHubRepoOpenError],
		})

		os.Exit(1)
	}
}

func HandlerOptionBuild() {
	packageManagerToUse := userInput.GetPackageManager()
	useCommitlint := userInput.AddCommitlint()
	shouldPublishToNpm := userInput.ShouldPublishToNpm()

	utils.CreateEmptyJsonFileIfNotExists(constants.PATH_PACKAGE_JSON)

	dependenciesToInstall := []string{"husky"}

	if useCommitlint {
		dependenciesToInstall = append(
			dependenciesToInstall,
			"lint-staged",
			"@commitlint/cli",
			"@commitlint/config-conventional",
		)
	}

	utils.InstallDependencies(utils.InstallProps{
		PackageManagerToUse: packageManagerToUse,
		PackagesToInstall:   dependenciesToInstall,
	})

	var wg sync.WaitGroup

	wg.Add(1)

	go func() {
		defer wg.Done()

		utils.GenerateHuskyConfig(
			utils.Props{
				PackageManagerToUse: packageManagerToUse,
				PackageJsonPath:     constants.PATH_PACKAGE_JSON,
				UseCommitlint:       useCommitlint,
				ShouldPublishToNpm:  shouldPublishToNpm,
			},
		)
	}()

	if useCommitlint {
		wg.Add(1)

		go func() {
			defer wg.Done()

			utils.GenerateCommitlintConfig(
				utils.CommitlintProps{
					PackageManagerToUse: packageManagerToUse,
					PackageJsonPath:     constants.PATH_PACKAGE_JSON,
					ShouldPublishToNpm:  shouldPublishToNpm,
				},
			)
		}()
	}

	wg.Wait()

	utils.WriteMessage(utils.WriteMessageProps{
		Type:    enums.MessageTypeSuccess,
		Message: "All tasks were completed",
	})
}
